articles/working-with-pythons-unittest/unittests.md

```diff
@@ -0,0 +1,221 @@
+2. Make sure you have Python3 installed. If not, install it from [here](https://www.python.org/downloads/).
+   Note that the Python standard library provides the `unittest` module hence no installation is required.
+3. Create two files: `mathfns.py` and `main.py` using the project structure shown below.
+
+```
+project-root
+    |
+    --- mathfns.py
+    --- main.py
+```
+
+### Example
+
+#### Program
+The `mathfns.py` file will store the units that we will perform unit testing on.
+The units are basic maths functions used to double a given number `x`, get the square of a number `x`, and to divide two numbers `num`(numerator) and `den`(denominator).
+
+```python
+# mathfns.py
+
+def double(x):
+   return 2 * x
+
+def square(x):
+   return x * x
+
+# divide numerator by denominator
+def divide(num, den):
+   return num / den
+```
+
+#### Unit tests
+The `main.py` file will store the unit tests that we have. We import the units `double`, `square`, and `divide` from the `mathfns` script defined above. We then import the `unittest` module from Python's standard library.
+
+The class `TestMathFunctions` inherits unit testing capabilities from the `unittest.TestCase` class. It provides three methods:
+1. `test_double` which does unit tests for the `double` unit
+2. `test_square` which does unit tests for the `square` unit
+3. `test_divide` which does unit tests for the `divide` unit
+
+The following naming convention is used to make the unit tests self-descriptive: the test names should be in the `test_UnitName` format. In this case, `UnitName` refers to the unit being tested.
+Inside each of the unit tests, we use assert methods to check for and report errors if any.
+
+```python
+# main.py
+
+from mathfns import double, square, divide
+import unittest
+
+class TestMathFunctions(unittest.TestCase):
+   def test_double(self):
+      # used to check that double(8) returns 16
+      self.assertEqual(double(8), 16)
+
+      # used to check that double(8) returns a number not equal to 15
+      self.assertNotEqual(double(8), 15)
+ 
+    def test_square(self):
+       # used to check that square(7) returns 49
+       self.assertEqual(square(7), 49)
+    
+    def test_divide(self):
+       # we want to make sure that the divide unit raises a ZeroDivisionError
+       # when one attempts to divide a number by zero.
+       with self.assertRaises(ZeroDivisionError):
+         divide(42, 0)
+
+if __name__ == '__main__':
+   # unittest.main() serves as the main entry point to run the unit tests.
+   # unit tests are not performed without this function call
+   unittest.main()
+```
+
+To run the unit tests, proceed as follows:
+```console
+>>> python3 -m unittest main.py
+```
+
+##### Output
+
+```console
+>>> python3 -m unittest main.py
+...
+----------------------------------------------------------------------
+Ran 3 tests in 0.000s
+
+OK
+```
+
+The above output shows that all three tests were executed correctly.
+
```

Comment:
Please delete the Output header and the last line. Then put:
The output will show that all three tests were executed correctly: